### PR TITLE
new bytecode emitter, GenBCode (5th attempt)

### PR DESCRIPTION
**Pull request by magarciaEPFL**
Compared to the previous submission this PR keeps line count reasonably small for all source files added.

This PR contains commits beyond those in the initial submission (that PR covered only the first two commits in this PR) with the aim of providing a better perspective on why the first two commits include the functionality they include. For example, dead-code elimination follows naturally bytecode generation, thus whenever GenBCode runs in the future it will perform dead-code elimination out of the box. That can be seen from commit "GenBCode tightens code before serializing classfiles"

Testing against the community build will provide more information than isolated tests.
Bytecode tests are by their very nature fragile (e.g., testing first that pattern matching emits a chain of jumps, only to make sure a moment later that chain of jumps has been collapsed by GenBCode). Concrete ideas and examples on how to balance usefulness vs brittleness are welcome.

Reviews about the content proper of the PR are most welcome.

review by @gkossakowski or @JamesIry or @retronym or @paulp

### CODE REVIEW

**Comment by magarciaEPFL** [comment]
Answering an offline question: nightly builds of scalac are no problem with the new backend / optimizer , under all optimization levels.

First, update build.xml to the desired optimization level:

```
-    <antcall target="@{name}"><param name="scalac.args.optimise" value="-optimise"/></antcall>
+    <antcall target="@{name}"><param name="scalac.args.optimise" value="-neo:o3"/></antcall>
```

Second, `ant all.clean && ant && ant replacelocker && ant nightly`

**Comment by magarciaEPFL** [comment]
Another time saving feature: to activate by default the new backend (thus making scalac 15% faster) simply change in `ScalaSettings`

```
  val neo = ChoiceSetting ("-neo", "choice of bytecode emitter", "Choice of bytecode emitter.",
                           List("GenASM", "GenBCode", "o1", "o2", "o3"),
                           "GenBCode") // in trunk "GenASM" should be the default
```

In fact, the default can be ticked up to `o1` (intra-method optimizations). The extra compilation time is negligible on, say, a two-core machine or better (negligible because those optimizations overlap in time with the compiler pipeline).

**Comment by magarciaEPFL** [comment]
The culprit of the failure above is `presentation/implicit-member`

Discussion at https://groups.google.com/d/msg/scala-internals/gYfPFunzooM/MypXEYEEGuwJ

File: src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+          }
+        }
+      }
+
+    /*
+     *  Checks for duplicate internal names case-insensitively,
+     *  builds ASM ClassNodes for mirror, plain, and bean classes;
+     *  enqueues them in queue-2.
+     *
+     */
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala:214]: Thanks for breaking this file up!

There are quite a few indentation problems around; here, above in `val caseInsensitively`. Could you please sweep over the patch to try to iron those out? I'll make time for more less superficial review (up to my knowledge to do so) in a few days.

**Pull request closed without merging**